src/subcommands/diff.rs

```diff
@@ -76,17 +99,97 @@ pub fn diff(
         .unwrap_or_else(|| {
             eprintln!("'{diff_bin}' process terminated without exit status.");
             config.error_exit_code
-        })
+        });
+
+    if code >= 2 {
+        let via_process_substitution =
+            |f: &Path| f.starts_with("/proc/self/fd/") || f.starts_with("/dev/fd/");
+        let is_git_diff = matches!(differ, Differ::GitDiff);
+        if is_git_diff
+            && code == 128
+            && (via_process_substitution(minus_file) || via_process_substitution(plus_file))
+        {
+            // https://stackoverflow.com/questions/22706714/why-does-git-diff-not-work-with-process-substitution
+            // When called as `delta <(echo foo) <(echo bar)`, then git < 2.42 just prints the diff of the
+            // filenames which were created by the process substitution and does not read their content.
+
+            // It looks like `git diff` failed due to lack of process substitution (version <2.42);
+            // try again with `diff`.
+            diff(minus_file, plus_file, Differ::Diff, config, writer)
```

Comment:
@th1000s I think this PR is almost there. I would like the following to be true:

> if a user has `git >= 2.42` then delta will use `git diff`, even if they use process substitution.

(So that their git config is honored and they can use all the `git diff` options).

Instead of trying to detect the user's git version, the current state of this PR always tries `git diff` and falls back to `diff` if it seems to have failed due to process substitution. After all, this problem should become rarer and rarer as people upgrade git.

This seems to be working on the two Mac laptops I tested on (one with old git and one new), since with the old git I am getting an _error_ (not merely printing of filenames):

```
$ git diff --no-index <(cat [b.rs](http://b.rs/)) src/[paint.rs](http://paint.rs/)
error: /dev/fd/11: unsupported file type
fatal: cannot hash /dev/fd/11
$ echo $?
128
$ git --version
git version 2.39.3 (Apple Git-146)
```

However it sounds like from your comment that you encountered incorrect output with a success (< 2) exit code:

> When called as `delta <(echo foo) <(echo bar)`, then git < 2.42 just prints the diff of the filenames which were created by the process substitution and does not read their content.

If so, then I think my PR needs to do a bit more.

Any thoughts? (and any chance you can test out this PR??)

Reply:
Indeed, on Linux git sees e.g. `/proc/self/fd/16`, then reads and diffs nonsense like `pipe:[15699060]`. But I would just document that this requires git v2.42 and not call and parse `git --version` first.

Reply:
> Indeed, on Linux git sees e.g. `/proc/self/fd/16`, then reads and diffs nonsense like `pipe:[15699060]`. But I would just document that this requires git v2.42 and not call and parse `git --version` first.

Just to make sure I understand, are you saying we should 

1. just not bother calling diff at all, now that git supports process substitution? If the user is trying process substitution then they should use a recent git that supports it.

or

2. that we should now always invoke git twice, i.e. the first time call git to parse the version and the second time call git or diff to compute the diff?

Reply:
I'd be okay with 1. because this will work everywhere.. eventually. But if 2. is not too much work now (and I remember that parsing `less --version` takes just 4ms) that would be nicer of course.

Reply:
Switched to parsing the git version. I think this PR is ready now.